### ISSUE TITLE
Remove package prefix from names, and provide init/update functions

### ISSUE THREAD

**Issue by QAFabianHuch** [aforemny/material-components-web-elm#76]
Currently, names have the package prefixed, and are meant to be exposed:
```elm
import Material.Button exposing (buttonConfig, textButton)

type Msg
    = ButtonClicked

main =
    textButton
        { buttonConfig | onClick = Just ButtonClicked }
        "Text"
```

However, this is not idiomatic, as names should generally not be exposed. Though without exposed values, structs cannot be updated (here: `buttonConfig`), so update functions for the fields had to be defined. With a few changes to argument order, this would look like:

```elm
import Material.Button as Button

type Msg
    = ButtonClicked
    | OtherMsg

main =
    Button.config
        |> Button.onClick ButtonClicked
        |> Button.textButton "Text"
```

However, this does have the disadvantage that fields can be set multiple times, which is nonsensical:
```elm
    Button.config
        |> Button.onClick ButtonClicked
        |> Button.onClick OtherMsg
        |> Button.textButton "Text"
```

I would still prefer the latter code style. What do other users think?

**Comment by aforemny** [comment]
Another benefit of the recommended (=latter) code style in contrast to the current one would be that the type `Button.Config` could be made opaque, and thus field additions to that record would not be a major version bump, but rather a minor version bump.

**Comment by aforemny** [comment]
Hi @alexanderkiel, @simonh1000, @bemyak, @jesterxl, @j340m3, since you are actively using this library, I would be curious about your opinion on this issue. Thanks!

**Comment by bemyak** [comment]
Thanks for reaching us, @aforemny !
I like this suggestion, it make API a bit cleaner. Also you made a good point about breaking changes: adding a new config option does not cause bump of a major version anymore.

**Comment by alexanderkiel** [comment]
@aforemny I'm with @QAFabianHuch here. Although that change would mean that you need one setter-function per config option, it would be way more idiomatic Elm and also play nicer with API stability. So I would opt for the suggested change.

**Comment by j340m3** [comment]
Thanks for asking, @aforemny! While yes, it would make it more ideomatic, and cleaner (big fan of pipelining!), I prefer the former one. 

1. As @QAFabianHuch pointed out, these functions need to take account of multiple calls. It's really hard to make these Impossible States impossible. What would be a good approach to assure consistency? Track all fields if they are set? And what would happen if this states are reached? 

```elm
type Button msg
    = TextButton ButtonConfig msg
    | Invalid
```
... doesn't feel right. What about 
```elm
config : Result ButtonError (ButtonConfig msg)
config =
    Ok defaultButtonConfig

onClick : msg -> Result ButtonError (ButtonConfig msg) -> Result ButtonError (Button msg)
onClick callback =
    -- Insert Has this-field-already-been-set logic
    Result.andThen (\x -> {x | onClick = callback})

textButton : String -> Result ButtonError (ButtonConfig msg) -> Button msg
textButton text config =
    Result.withDefault defaultButtonConfig
    |> TextButton text
```
... I believe its the standard trade-off between safety and convenience. For my part, I like to eliminate the Impossible States, makes me worry about one thing less. I like the compiler to hammer on my fingers each time I make a mistake. When the code compiles, I can be sure that it has a certain level of quality.
2. With this interface, we make it harder to reason about buttons. Today I can do
```elm

appendRelativeIfEmpty : { a | additionalAttributes : List (Attributes msg) } -> { a | additionalAttributes :  List (Attributes msg)}
appendRelativeIfEmpty config =
     if List.len config.additionalAttributes > 0 then
          config
     else
          { config | additionalAttributes = [Html.Attributes.style "position" "relative"]
```
This could be done by exposing getters (or updater) as well as the aforementioned (@alexanderkiel) setters, but might also look ugly. 

**TLDR**:This is only my gut feeling. I like the iomatic way of chaining Monads, but I fear the vagueness and the loss of expressiveness.

**Comment by alexanderkiel** [comment]
@j340m3 I don't see that this multiple function calls setting record files are a problem. While it isn't possible that you set the same record field twice inside one record syntax, it's perfectly possible in a two step process. That's why it's possible to overwrite the default config returned currently. Also setting a field twice doesn't make the config invalid. You just set a field twice and the last setting will be used.

On the other hand, your proposed double-setting error handling approach would lead to somehow unexpected use of the default config in case you accidentally set a field twice.

I don't see how the compiler could help us here. But please correct me.

**Comment by j340m3** [comment]
@alexanderkiel Yes, that multiple call was not thought until the end and more a vague sketch. You're absolutely right. So I'm in as longs as there will be getters, too :wink:

**Comment by aforemny** [comment]
Hi, I have been recently looking at this and I just asked about one thing that I am unsure about on Slack. I thought I paste the question here, so that I can be sure to get your answers as well:
>
> Hi! When comparing the builder pattern to record based configuration, I am assuming that a record `type alias Config = { disabled : Bool }` would be mirrored by `setDisabled : Config -> Config`; note the "loss" of the `Bool` argument. I am facing a problem where for some builder-style functions I like that loss and for some I don't and for some others I am unsure. I am imagining that some functions are relatively static so that it does not make sense to parameterize them: either they are always used with True as parameter, or they are not used at all. Some of them I anticipate to be dynamic and so it makes sense to parameterize them. The hard part is the category of functions that I am not sure about. Should I parameterize them or not?
>
> Maybe a good example for static functions would be `setPrimaryColor` for a primary action button. I think `setDisabled` is a good candidate for a dynamic function because whether an input is disabled depends on what the user enters and so it is expected to change states. A good example for a function that I am not sure about could be `setDense` that makes the margins around a UI component smaller than normally. I would expect this to be used statically, but what if the user has a button that toggles between dense and non-dense styling of the page?
>
> I have also thought about the followng alternatives:
>
> - make all setters static, and offer dynamic setters for the ones where I expect dynamic use like `setDisabledIf`
> - make all setters dynamic, ie. `setPrimaryColor : Bool -> Config -> Config`
> - make all setters static, and offer utility functions: `ifTrue : Bool -> (Config -> Config) -> (Bool -> Config -> Config)`, `ifJust : Maybe a -> (a -> Config -> Config) -> (Maybe a -> Config -> Config)`, etc.
> - make all setters static, and do not offer utility functions; expect the user to implement whatever variant `ifTrue` and `ifJust` themselves
>
> I would be curious about your thoughts! Do you know of packages that I could look at to see how they solved that or maybe a similar problem? Thanks for your feedback!

I hope you don't mind me pinging everyone that I think is interested in this: @QAFabianHuch @simonh1000 @bemyak @j340m3 @alexanderkiel @matsjoyce 

Thanks for your feedback!

**Comment by matsjoyce** [comment]
IMO the "dynamic for everything" style is the cleanest. The static style lacks flexibility, and adding helper functions to work around the lack of flexibility just makes things more complicated. Having a mixed style means that the programmer has to remember which are parameterized and which aren't, and would still require helpers for the static cases.

As a data point, I tend to use `disabled` like this (`m.sessions` is populated by a HTTP request):

```elm
Button.raisedButton { buttonConfig | onClick = Just SelectSession, disabled = m.sessions == Nothing } "Change Session"
```

A quick grep of my code base shows that none of my uses of `disabled` are static, and so even methods which should obviously be static (like `setPrimaryColor`) probably aren't static for someone. Adding `True` isn't that much more typing. And the consistency looks nicer (IMO):

```elm
Button
|> setDisabled True
|> setPrimaryColor True
|> setText "Hai"
|> setOnClick Clicked
```

Vs:

```elm
Button
|> setDisabled
|> setPrimaryColor
|> setText "Hai"
|> setOnClick Clicked
```

For other packages:
 - Bootstrap uses dynamic for `disabled`, static for the button colors (but Bootstrap has a lot of button colors): https://package.elm-lang.org/packages/rundis/elm-bootstrap/latest/Bootstrap-Button#disabled
 - MDL uses static for pretty much everything: https://package.elm-lang.org/packages/debois/elm-mdl/latest/Material-Button#disabled
 - The HTML module uses dynamic: https://package.elm-lang.org/packages/elm/html/latest/Html-Attributes#disabled

**Comment by alexanderkiel** [comment]
I would also vote for dynamic for everything.

**Comment by aforemny** [comment]
I have progressed on this in the builder-pattern branch #104. There is still some work to be done, namely updating the demo and updating the documentation in places. But I am expecting to merge that branch into a new release shortly.

If anyone would be willing to update to the new API before then, I would appreciate any feedback. Thanks!

**Comment by alexanderkiel** [comment]
Thanks I currently refactor my application. I have a first question: For example in TextField, there are some setters which use Maybe types like `setPlaceholder` and some which don't like `setOnInput`. What is the reasoning behind doing one or the other?

**Comment by alexanderkiel** [comment]
If someone is interested in the amount of changes required, my pull request for my own project is here: https://github.com/samply/blaze-quality-reporting-ui/pull/5

I used the following function to lift setters without a Maybe type into setter with a Maybe type:

```elm
liftMaybe : (a -> config -> config) -> (Maybe a -> config -> config)
liftMaybe setter =
    \maybeValue config ->
        case maybeValue of
            Just value ->
                setter value config

            Nothing ->
                config
```

With that I can set Button onClick handlers this way:

```elm
Button.config
    |> liftMaybe Button.setOnClick maybeOnClick
```

So maybe it would be a good idea to remove the Maybe types from all setters. Except if one likes to pre-build a config and needs to unset a certain setting. In this case we would need unset functions.

**Comment by aforemny** [comment]
Hi @alexanderkiel, thank you for testing my changes! The idea was to not have functions `liftMaybe` to keep the API surface of this library smaller. Because of that, most configuration options do have a `Bool`, or a `Maybe` argument.

I have decided to not use maybes in some places, though, most notably in event handler configuration options. I am surprised you have a `maybeOnClick` for a button. Why is that? Thanks!

**Comment by alexanderkiel** [comment]
>  I am surprised you have a maybeOnClick for a button. Why is that? Thanks!

I use `Maybe` messages in situations where I have messages carrying data which changes and can be unknown. However I can also put this data in my model because I have to put this data carrying messages into my model anyway. As I coded it this way, I thought it would be a good idea to have everything I need to process this kind of messages inside the message itself, but if I think about it in more depth, it's not really an advantage.

In general, I think you are right that messages should be just static. For static messages, I see no need for `Maybe`.

**Comment by aforemny** [comment]
Implemented in 96ee3e7ba5f8a208fb4500e32efc0ce7b807220b. Thanks for everyone's input on this!

**Comment by alexanderkiel** [comment]
Thanks for this huge amount of work!
